### ISSUE TITLE
Asynchronous Validation

### ISSUE THREAD

**Issue by NickMitrokhin** [DevExpress/DevExtreme#9319]
# The Problem

The DevExtreme validation API does not support async calls and cannot be used to implement server-side validation. You can apply the solution described in the [Server-Side Validation](https://js.devexpress.com/Documentation/Guide/Widgets/Common/UI_Widgets/Data_Validation/#Server-Side_Validation) article, but it does not allow displaying the current validation status in the editor being validated.

# The Proposed Solution

We plan to add a new `async` validation rule whose API is similar to the [custom validation rule](https://js.devexpress.com/Documentation/ApiReference/UI_Widgets/dxValidator/Validation_Rules/CustomRule/)'s API. The only difference is that the [validationCallback](https://js.devexpress.com/Documentation/ApiReference/UI_Widgets/dxValidator/Validation_Rules/CustomRule/#validationCallback) function should return the [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) object. 

## Validation API

Add the `async` rule to the [dxValidator.validationRules](https://js.devexpress.com/Documentation/ApiReference/UI_Widgets/dxValidator/Configuration/#validationRules) array:

```js
validationRules: [
    ...,
    {
        type: "async",
        message: "Value is invalid",
        validationCallback: function(params) {
            return new Promise(...);
        }
    }
]
```

Fulfill the Promise with **true** if the value is valid or **false** otherwise. If the server sends the validation error message, fulfill the Promise with the following object:

```js
{
    isValid: false,
    message: 'Your error message'
}
```

The Promise can also be rejected with an error. The dxValidator displays all the error messages.

![invalid](https://user-images.githubusercontent.com/1420883/63158117-3e1c3300-c021-11e9-9057-0a3f89e91fd3.png)

## Rule-processing order

All synchronous rules are checked first. The checks are run rule by rule in the sequence the rules have in the **validationRules** array. If these checks are successfull, the validation engine checks all asynchronous rules simultaneously. 

We also plan to add a **status** option to the [validation result](https://js.devexpress.com/Documentation/ApiReference/UI_Widgets/dxValidator/Validation_Result/) that shows the status of validation requests:

- "PENDING" - At least one async rule is not fulfilled or rejected yet.
- "VALID" - No validation rules were broken.
- "INVALID" - At least one validation rule was broken.

While the **status** is "PENDING", the editor displays a loading indicator:

![pending](https://user-images.githubusercontent.com/1420883/63158118-3e1c3300-c021-11e9-9e35-86deb4a721be.gif)

# Try It
## Live Sandbox
[jQuery](https://codepen.io/nickmitrokhin/pen/XWWWLML?editors=0010)
[Angular](https://codesandbox.io/s/overview-devextreme-validation-t3un4)
[React](https://codesandbox.io/s/overview-devextreme-validation-6npvb)
[Vue](https://codesandbox.io/s/overview-devextreme-validation-4sywl)

## Installation
npm install devextreme@19.2.2-beta

# We need Your Feedback

## Take a Quick Poll
[Do the new asynchronous validation features meet your needs?](https://forms.gle/y9kik64nRtbHgx7y9)

**Comment by NickMitrokhin** [comment]
A new DevExtreme version 19.2-beta is available now. You can test this feature using examples that reside in the Live Sandbox section of the discussion page.

**Comment by Bykiev** [comment]
@NickMitrokhin, can you please publish this update on bower too?

**Comment by Bykiev** [comment]
@NickMitrokhin, I've one qustion about the validation, but not fully related to async validation. My goal is validating the form editors against some rules on the server side. To make 1 universal method on server side for each form dataField I need to pass property name in this method. There is no other way except defining own callback function for each property I need to validate and passing dataField name on backend. I've created support ticket (T821472), but still didn't get any desirable solution... I think the best way will be extending form validationRules with dataField name. What do you think about it?

**Comment by NickMitrokhin** [comment]
>can you please publish this update on bower too?

As you know Bower is deprecated and even Bower creators recommend using any other package managers. That is why we did not publish the Beta version using Bower. Would you please elaborate on why you cannot use NPM in your case? Please describe the issue in greater detail.

**Comment by Bykiev** [comment]
@NickMitrokhin, we're aware of bower was deprecated, I don't see any problems to migrate on npm. The only reason we're still using bower because it's easy to manage bower packages in VS 2017 same way as NuGet. Thanks!

**Comment by NickMitrokhin** [comment]
Thank you for your clarification. At the moment, we do not have plans to publish DevExtreme 19.2 on Bower. So, I recommend that you use NPM in your case.

**Comment by Bykiev** [comment]
@NickMitrokhin, how to remove green mark on success validation?

**Comment by San4es** [comment]
@Bykiev 
Currently, it is possible only via overriding CSS. Would you share with us your scenario where this may be needed?

**Comment by Bykiev** [comment]
@San4es, I don't like that fields with async validation rules has green marks, but other validation rules doesn't have it. From UI side they should be looking the same way, I think.

**Comment by San4es** [comment]
Thank you for the clarification. Your vote is counted :) To make a decision on adding an API for displaying the validation mark, we will collect additional feedback from our customers regarding their cases and requirements.

**Comment by steelcg** [comment]
A good solution for async validation, but why that isn't supported in DxDataGrid whit batch edit mode?
E1057 - The dxDataGrid widget does not support asynchronous validation rules in the "batch" edit mode. See:
http://js.devexpress.com/error/19_2/E1057

**Comment by NickMitrokhin** [comment]
@steelcg,

Async validation is not supported for the Batch, Cell, and Row editing modes because dxDataGrid requires significant changes in its implementation. We are planning to support "async" validation rules for these modes in one of the next major versions.

**Comment by NickMitrokhin** [comment]
Starting with 20.1, async validation will be supported in DataGrid for Row, Batch and Cell editing modes. You can test it in v20.1.2-alpha. To install this version, execute the following command:

```
npm install devextreme@20.1.2-alpha-20064-2016
```

**Comment by San4es** [comment]
Thanks to everybody who gave us early feedback on this feature. The feature is now available in the [v19.2 release](https://github.com/DevExpress/DevExtreme/releases/tag/19.2.3) (for editors and Form), and in the [v20.1 release](https://github.com/DevExpress/DevExtreme/releases/tag/20.1.3) (for Data Grid and Tree List). I'm closing this thread. In the case of bugs or questions, feel free to create new [GitHub issues](https://github.com/DevExpress/DevExtreme/issues) or tickets in our [Support Center](https://www.devexpress.com/Support/Center/).
